modules/raft/src/main/java/org/apache/ignite/internal/raft/storage/impl/VaultGroupStoragesDestructionIntents.java

```diff
@@ -0,0 +1,146 @@
+    private static final ByteOrder BYTE_UTILS_BYTE_ORDER = ByteOrder.BIG_ENDIAN;
+
+    private static final String PARTITION_GROUP_NAME = "partition";
+
+    private static final int RAFT_GROUPS = 3;
+
+    private final VaultManager vault;
+
+    private final ConcurrentMap<String, RaftGroupOptionsConfigurer> configurerByName = new ConcurrentHashMap<>();
+
+    /** Constructor. */
+    public VaultGroupStoragesDestructionIntents(VaultManager vault) {
+        this.vault = vault;
+    }
+
+    @Override
+    public void addGroupOptionsConfigurer(ReplicationGroupId groupId, RaftGroupOptionsConfigurer groupOptionsConfigurer) {
+        configurerByName.put(groupId.toString(), groupOptionsConfigurer);
+    }
+
+    @Override
+    public void addPartitionGroupOptionsConfigurer(RaftGroupOptionsConfigurer groupOptionsConfigurer) {
+        configurerByName.put(PARTITION_GROUP_NAME, groupOptionsConfigurer);
+    }
+
+    @Override
+    public void saveDestroyStorageIntent(RaftNodeId nodeId, RaftGroupOptions groupOptions) {
+        String configurerName = nodeId.groupId() instanceof PartitionGroupId ? PARTITION_GROUP_NAME : nodeId.groupId().toString();
+
+        vault.put(
+                buildKey(nodeId.nodeIdStringForStorage()),
+                toBytes(new DestroyStorageIntent(configurerName, groupOptions.volatileStores()))
+        );
+    }
+
+    @Override
+    public void removeDestroyStorageIntent(String nodeId) {
+        vault.remove(buildKey(nodeId));
+    }
+
+    private static ByteArray buildKey(String nodeId) {
+        byte[] nodeIdBytes = nodeId.getBytes(UTF_8);
+
+        byte[] key = ByteBuffer.allocate(GROUP_STORAGE_DESTRUCTION_PREFIX.length + nodeIdBytes.length)
+                .order(BYTE_UTILS_BYTE_ORDER)
+                .put(GROUP_STORAGE_DESTRUCTION_PREFIX)
+                .put(nodeIdBytes)
+                .array();
+
+        return new ByteArray(key);
+    }
+
+    private static String raftNodeIdFromKey(byte[] key) {
+        return new String(key, GROUP_STORAGE_DESTRUCTION_PREFIX.length, key.length - GROUP_STORAGE_DESTRUCTION_PREFIX.length, UTF_8);
+    }
+
+    @Override
+    public Map<String, RaftGroupOptions> readGroupOptionsByNodeIdForDestruction() {
+        assert configurerByName.size() == RAFT_GROUPS
+                : "Configurers for CMG, metastorage and partitions must be added, got: " + configurerByName.keySet();
+
+        try (Cursor<VaultEntry> cursor = vault.prefix(new ByteArray(GROUP_STORAGE_DESTRUCTION_PREFIX))) {
+            Map<String, RaftGroupOptions> result = new HashMap<>();
+
+            while (cursor.hasNext()) {
+                VaultEntry next = cursor.next();
+
+                String nodeId = raftNodeIdFromKey(next.key().bytes());
+
+                // todo add serializer
+                DestroyStorageIntent intent = ByteUtils.fromBytes(next.value());
+
+                RaftGroupOptions groupOptions = intent.isVolatile
```

Comment:
I don't like that we build group options just to conform to the 'group options configurer' abstraction, we are making a loop.. But mybe we should change this later (probably with revisiting the 'group options configurer' abstraction itself)